crates/red_knot_python_semantic/src/visibility_constraints.rs

```diff
@@ -0,0 +1,326 @@
+//!         x = 2
+//!
+//! use(x)
+//! ```
+//! At the usage of `x`, i.e. after control flow has been merged again, the visibility of the `x =
+//! 0` binding behaves as follows: the binding is always visible if `test1` is always-false *or*
+//! `test2` is always-false; and it is never visible if `test1` is always-true *and* `test2` is
+//! always-true. This corresponds to a ternary *OR* operation in Kleene logic:
+//!
+//!       | OR           | always-false | ambiguous    | always-true  |
+//!       |--------------|--------------|--------------|--------------|
+//!       | always false | always-false | ambiguous    | always-true  |
+//!       | ambiguous    | ambiguous    | ambiguous    | always-true  |
+//!       | always true  | always-true  | always-true  | always-true  |
+//!
+//! Using this, we can annotate the visibility constraints for the example above:
+//! ```py
+//! x = 0  # ~test1 OR ~test2
+//!
+//! if <…>:
+//!     if test1:
+//!         x = 1  # test1
+//! else:
+//!     if test2:
+//!         x = 2  # test2
+//!
+//! use(x)
+//! ```
+//!
+//! ### Explicit ambiguity
+//!
+//! In some cases, we explicitly add a `VisibilityConstraint::Ambiguous` constraint to all bindings
+//! in a certain control flow path. We do this when branching on something that we can not (or
+//! intentionally do not want to) analyze statically. `for` loops are one example:
+//! ```py
+//! x = <unbound>
+//!
+//! for _ in range(2):
+//!    x = 1
+//! ```
+//! Here, we report an ambiguous visibility constraint before branching off. If we don't do this,
+//! the `x = <unbound>` binding would be considered unconditionally visible in the no-loop case.
+//! And since the other branch does not have the live `x = <unbound>` binding, we would incorrectly
+//! create a state where the `x = <unbound>` binding is always visible.
+//!
+//!
+//! ### Properties
+//!
+//! The ternary `AND` and `OR` operations have the property that `~a OR ~b = ~(a AND b)`. This
+//! means we could, in principle, get rid of either of these two to simplify the representation.
+//!
+//! However, we already apply negative constraints `~test1` and `~test2` to the "branches not
+//! taken" in the example above. This means that the tree-representation `~test1 OR ~test2` is much
+//! cheaper/shallower than basically creating `~(~(~test1) AND ~(~test2))`. Similarly, if we wanted
+//! to get rid of `AND`, we would also have to create additional nodes. So for performance reasons,
+//! there is a small "duplication" in the code between those two constraint types.
+//!
+//! [Kleene]: <https://en.wikipedia.org/wiki/Three-valued_logic#Kleene_and_Priest_logics>
+
+use ruff_index::IndexVec;
+
+use crate::semantic_index::ScopedVisibilityConstraintId;
+use crate::semantic_index::{
+    ast_ids::HasScopedExpressionId,
+    constraint::{Constraint, ConstraintNode, PatternConstraintKind},
+};
+use crate::types::{infer_expression_types, Truthiness};
+use crate::Db;
+
+const MAX_RECURSION_DEPTH: usize = 24;
```

Comment:
Maybe add a comment here recording your performance experimentation on the Black codebase that led to picking this number?